src/core/mount.c

```diff
@@ -1280,6 +1281,22 @@ static void mount_sigchld_event(Unit *u, pid_t pid, int code, int status) {
         if (pid != m->control_pid)
                 return;
 
+        /* So here's the thing, we really want to know before /usr/bin/mount or /usr/bin/umount exit whether
+         * they established/remove a mount. This is important when mounting, but even more so when unmounting
+         * since we need to deal with nested mounts and otherwise cannot safely determine whether to repeat
+         * the unmounts. In theory, the kernel fires /proc/self/mountinfo changes off before returning from
+         * the mount() or umount() syscalls, and thus we should see the changes to the proc file before we
+         * process the waitid() for the /usr/bin/(u)mount processes. However, this is unfortunately racy: we
+         * have to waitid() for processes using P_ALL (since we need to reap unexpected children that got
+         * reparented to PID 1), but when using P_ALL we might end up reaping processes that terminated just
+         * instants ago, after our last event loop iteration where we might have noticed /proc/self/mountinfo
```

Comment:
This should probably read "where we might have not noticed". I mean if we noticed the change before dispatching SIGCHLD we wouldn't have to recheck it here right?

Reply:
well, i wanted to say with that that we only see the changes via epoll, and that was in the last iteration, which was the last point in time where we could have seen such an event.

will reword slightly, to make this a bit clearer.